paddle/framework/net_design.md

```diff
@@ -0,0 +1,250 @@
+# Network Design
+
+`Network` is the container and controller of a set of operators,
+users can build a real network from a `NetDef` which is a protobuf message 
+and use `Network.Run()` to run all the operators in the network.
+
+The `Network` will
+
+- manage all the operators contained in the network.
+- not own any `Variable`.
+
+# API
+
+## NetBase
+To make the `Network` extendable, a base class is defined like this
+
+```c++
+// operator's index stored in a network.
+typedef int OpIndex;
+
+// The minimum a network should be implemented.
+class NetBase {
+ public:
+  // run all the operators and return success(true) or not, all the
+  // variables are located in `scope`. `ctx` describes the detail execution
+  // environment for ops. `begin` and `end` specify the scope of `ops_` to run,
+  // If no positive indexes are provided, all operators in `ops_` will run.
+  virtual bool Run(Scope *scope, OpContext *ctx, OpIndex begin = -1,
+                   OpIndex end = -1) const = 0;
+
+  // Add an Operator according to `def`.
+  virtual OpIndex AddOp(const proto::OpDef &def) = 0;
+
+  // Add optimizer operators acctording to `attrs`.
+  virtual void AddOptimizerOps(const OptAttrs &attrs) = 0;
+
+  // Add backward operators.
+  virtual void AddBackwardOps() = 0;
+
+  // Infer the shapes of variables required by operators in the network. The
+  // `scope` will be mutated according to the inferred shapes.
+  virtual bool InferShape(Scope *scope) = 0;
```

Comment:
Or we can remove this function?  Each operator has its own `InferShape', just call `op.InferShape` in `BaseNet.AddOp`.  Make the `AddOp` the only method to add an operator to a network, so that `op.InferShape` will be called every time automatically, no need to call  `BaseNet.InferShape` manually.